### PR TITLE
added algorithm for binary exponent

### PR DESCRIPTION
**Pull request by NEERAJAP2001**
### **Describe your change:**

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: maths/binary_exponent.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    True
+    """
+    result = 1
+    while expo:
+        if expo % 2:
+            result *= num
+            expo = expo - 1
+        else:
+            num *= num
+            expo /= 2
```
Comment (cclauss) [inline maths/binary_exponent.py:17]:
```suggestion
        div, mod = divmod(expo, 2)
        if mod:
            result *= num
            expo -= 1
        else:
            num *= num
            expo = div
```

File: maths/binary_exponent.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
+def bin_exp(num, expo) -> int:
+    """
+    This function calculates the exponent of 'num' over 'expo' and
+    returns a positive number.
```
Comment (cclauss) [inline maths/binary_exponent.py:4]:
```suggestion
    Calculate the exponent of 'num' over 'expo' and return a positive number.
```

File: maths/binary_exponent.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    This function does binary exponential hence O(log N) time complexity.
+    So, It is faster than normal exponential function for huge numbers.
+    >>> bin_exp(2,5)==32
+    True
```
[the remark this answers is not in the document]
Comment (cclauss) [reply maths/binary_exponent.py:8]: Need more doctests: (0, 0), (-2, 0), (0, 2), (-2, -2), (1.5, 2.5), ('a', 'b'), (5, None)

File: maths/binary_exponent.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+def bin_exp(num, expo) -> int:
```
Comment (cclauss) [inline maths/binary_exponent.py:1]:
```suggestion
def bin_exp(num: int, expo: int) -> int:
```

**Review by cclauss — CHANGES_REQUESTED** [review]

File: maths/binary_exponent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NEERAJAP2001) [reply maths/binary_exponent.py]: okay sure, sir!

File: maths/binary_exponent.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+def bin_exp(num: int, expo: int) -> int:
+    """
+    Calculate the exponent of 'num' over 'expo' and return a positive number.
+    This function does binary exponential hence O(log N) time complexity.
+    So, It is faster than normal exponential function for huge numbers.
+    >>> bin_exp(2, 5)
+    32
+    >>> bin_exp(1,0)
```
Comment (cclauss) [inline maths/binary_exponent.py:8]:
```suggestion
    >>> bin_exp(1, 0)
```

File: maths/binary_exponent.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    1
+    >>> bin_exp(-2, -2)
+    0
+    >>> bin_expo(1.5, 2.5)
+    2
+    >>> bin_expo('a', 'b')
+    TypeError: can't multiply sequence by non-int of type 'str'
+    >>> bin_expo(5, None)
+    TypeError: can't multiply sequence by non-int of type 'str'
+
```
Comment (cclauss) [inline maths/binary_exponent.py:20]:
```suggestion
```

File: maths/binary_exponent.py [hunk 7]
```diff
@@ -11,24 +12,28 @@ def bin_exp(num: int, expo: int) -> int:
...
+        try:
+            div, mod = divmod(expo, 2)
+            if mod:
+                result *= num
+                expo -= 1
+            else:
+                num *= num
+                expo = div 
+            return result  
+        except:
```
Comment (cclauss) [inline maths/binary_exponent.py]: Bare exceptions are not accepted in this repo.  https://realpython.com/the-most-diabolical-python-antipattern

**Pull request closed without merging**